ProcessLib/PhaseField/PhaseFieldFEM.h

```diff
@@ -0,0 +1,671 @@
+                                                        phasefield_size> Kud;
+        Kud.setZero(displacement_size, phasefield_size);
+
+        typename ShapeMatricesType::template MatrixType<phasefield_size,
+                                                        displacement_size> Kdu;
+        Kdu.setZero(phasefield_size, displacement_size);
+
+        typename ShapeMatricesType::NodalMatrixType Kdd;
+        Kdd.setZero(phasefield_size, phasefield_size);
+
+        typename ShapeMatricesType::NodalMatrixType Ddd;
+        Ddd.setZero(phasefield_size, phasefield_size);
+
+        double const& dt = _process_data.dt;
+
+        SpatialPosition x_position;
+        x_position.setElementID(_element.getID());
+
+        unsigned const n_integration_points =
+            _integration_method.getNumberOfPoints();
+        for (unsigned ip = 0; ip < n_integration_points; ip++)
+        {
+            x_position.setIntegrationPoint(ip);
+            auto const& w = _ip_data[ip].integration_weight;
+
+            auto const& dNdx = _ip_data[ip].dNdx;
+            auto const& N = _ip_data[ip].N;
+
+            auto const x_coord =
+                interpolateXCoordinate<ShapeFunction, ShapeMatricesType>(
+                    _element, N);
+            auto const& B = LinearBMatrix::computeBMatrix<
+                DisplacementDim, ShapeFunction::NPOINTS,
+                typename BMatricesType::BMatrixType>(dNdx, N, x_coord,
+                                                     _is_axially_symmetric);
+
+            auto const& C_tensile = _ip_data[ip].C_tensile;
+            auto const& C_compressive = _ip_data[ip].C_compressive;
+
+            auto& eps = _ip_data[ip].eps;
+            auto const& strain_energy_tensile =
+                _ip_data[ip].strain_energy_tensile;
+            auto const& sigma_tensile = _ip_data[ip].sigma_tensile;
+
+            auto& history_variable = _ip_data[ip].history_variable;
+            auto& history_variable_prev = _ip_data[ip].history_variable_prev;
+
+            auto const& sigma_real = _ip_data[ip].sigma_real;
+
+            // Kdd_1 defines one term which both used in Kdd and local_rhs for
+            // phase field
+            double const gc = _process_data.crack_resistance(t, x_position)[0];
+            double const ls =
+                _process_data.crack_length_scale(t, x_position)[0];
+            typename ShapeMatricesType::NodalMatrixType const Kdd_1 =
+                dNdx.transpose() * 2 * gc * ls * dNdx;
+
+            //
+            // displacement equation, displacement part
+            //
+            double const k = _process_data.residual_stiffness(t, x_position)[0];
+            double const d_ip = N.dot(d);
+            double const degradation = d_ip * d_ip * (1 - k) + k;
+            eps.noalias() = B * u;
+            auto C = _ip_data[ip].updateConstitutiveRelation(t, x_position, dt, u,
+                                                    degradation);
```

Comment:
There is a warning about the `C` variable being unused. I think the call to `material_model.integrateStress(...)` is unnecessary, because everything is handled in the `material_model.calculateDegradedStress(...)` function.